File: languages/go/tests/enforcement_test.go
package oso_test

import (
	"fmt"
	"reflect"
	"testing"

	oso "github.com/osohq/go-oso"
	"github.com/osohq/go-oso/errors"
	"github.com/osohq/go-oso/types"
)

type Request struct {
	Method string
	Path   string
}

func getOso(t *testing.T) oso.Oso {
	var o oso.Oso
	var err error
	if o, err = oso.NewOso(); err != nil {
		t.Fatalf("Failed to set up Oso: %v", err)
	}

	o.RegisterClassWithNameAndFields(reflect.TypeOf(User{}), nil, "User", map[string]interface{}{
		"Name": "String",
	})
	o.RegisterClassWithNameAndFields(reflect.TypeOf(Widget{}), nil, "Widget", map[string]interface{}{
		"Id": "Integer",
		"Parent": types.Relation{
			Kind:       "one",
			OtherType:  "Company",
			MyField:    "CompanyId",
			OtherField: "Id",
		},
	})
	o.RegisterClassWithNameAndFields(reflect.TypeOf(Company{}), nil, "Company", map[string]interface{}{
		"Id": "Integer",
	})
	o.RegisterClassWithNameAndFields(reflect.TypeOf(Request{}), nil, "Request", map[string]interface{}{
		"Method": "String",
		"Path":   "String",
	})

	return o
}

func assertAuthorizationError(t *testing.T, err error, isNotFound bool) {
	if err == nil {
		t.Fatal("Expected forbidden error from Authorize")
	}
	switch err.(type) {
	case *errors.NotFoundError:
		if !isNotFound {
			t.Error("Expected ForbiddenError, got NotFoundError")
		}
	case *errors.ForbiddenError:
		if isNotFound {
			t.Error("Expected NotFoundError, got ForbiddenError")
		}
	default:
		t.Errorf("Unexpected error from Authorize: %v", err)
	}
}

func assertSetEqual(t *testing.T, results map[interface{}]struct{}, elements []string) {
	if len(results) != len(elements) {
		t.Errorf("Expected %v to contain exactly %v", results, elements)
	}
	for _, element := range elements {
		if _, ok := results[element]; !ok {
			t.Errorf("Expected to find %v in %v", element, results)
		}
	}
}

func TestAuthorize(t *testing.T) {
	o := getOso(t)
	var err error

	guest := User{Name: "guest"}
	admin := User{Name: "admin"}
	widget0 := Widget{Id: 0}
	widget1 := Widget{Id: 1}

	if err = o.LoadString("allow(_actor: User, \"read\", widget: Widget) if " +
		"widget.Id = 0; " +
		"allow(actor: User, \"update\", _widget: Widget) if " +
		"actor.Name = \"admin\";"); err != nil {

		t.Errorf("LoadString returned error: %v", err)
	}

	if err = o.Authorize(guest, "read", widget0); err != nil {
		t.Errorf("Authorize returned error for allowed action: %v", err)
	}
	if err = o.Authorize(admin, "update", widget1); err != nil {
		t.Errorf("Authorize returned error for allowed action: %v", err)
	}

	// Throws a forbidden error when user can read resource
	err = o.Authorize(guest, "update", widget0)
	assertAuthorizationError(t, err, false)

	// Throws a not found error when user cannot read resource
	err = o.Authorize(guest, "read", widget1)
	assertAuthorizationError(t, err, true)
	err = o.Authorize(guest, "update", widget1)
	assertAuthorizationError(t, err, true)
}

func TestAuthorizeRequest(t *testing.T) {
	o := getOso(t)
	var err error

	guest := User{Name: "guest"}
	verified := User{Name: "verified"}

	if err = o.LoadString("allow_request(_: User{Name: \"guest\"}, request: Request) if " +
		"  request.Path = \"/repos\"; " +
		"allow_request(_: User{Name: \"verified\"}, request: Request) if " +
		"  request.Path = \"/account\"; "); err != nil {

		t.Errorf("LoadString returned error: %v", err)
	}

	if err = o.AuthorizeRequest(guest, Request{"GET", "/repos"}); err != nil {
		t.Errorf("Authorize returned error for allowed action: %v", err)
	}
	err = o.AuthorizeRequest(guest, Request{"GET", "/other"})
	assertAuthorizationError(t, err, false)

	if err = o.AuthorizeRequest(verified, Request{"GET", "/account"}); err != nil {
		t.Errorf("Authorize returned error for allowed action: %v", err)
	}
	err = o.AuthorizeRequest(guest, Request{"GET", "/account"})
	assertAuthorizationError(t, err, false)
}

func TestAuthorizeField(t *testing.T) {
	o := getOso(t)
	var err error

	o.LoadString( // Admins can update all fields
		"allow_field(actor: User, \"update\", _widget: Widget, field) if " +
			"  actor.Name = \"admin\" and " +
			"  field in [\"name\", \"purpose\", \"private_field\"]; " +

			// Anybody who can update a field can also read it
			"allow_field(actor, \"read\", widget: Widget, field) if " +
			"  allow_field(actor, \"update\", widget, field); " +

			// Anybody can read public fields
			"allow_field(_: User, \"read\", _: Widget, field) if " +
			"  field in [\"name\", \"purpose\"];")

	admin := User{"admin"}
	guest := User{"guest"}
	widget := Widget{0, 0}

	if err = o.AuthorizeField(admin, "update", widget, "purpose"); err != nil {
		t.Errorf("Authorize returned error for allowed action: %v", err)
	}
	err = o.AuthorizeField(admin, "update", widget, "foo")
	assertAuthorizationError(t, err, false)

	if err = o.AuthorizeField(guest, "read", widget, "purpose"); err != nil {
		t.Errorf("Authorize returned error for allowed action: %v", err)
	}
	err = o.AuthorizeField(guest, "read", widget, "private_field")
	assertAuthorizationError(t, err, false)
}

func TestAuthorizedActions(t *testing.T) {
	o := getOso(t)
	var err error

	o.LoadString("allow(_actor: User{Name: \"Sally\"}, action, _resource: Widget{Id: 1}) if action in [\"CREATE\", \"READ\"];")

	actor := User{Name: "Sally"}
	resource := Widget{Id: 1}

	res, err := o.AuthorizedActions(actor, resource, false)
	if err != nil {
		t.Fatalf("Failed to get allowed actions: %v", err)
	}
	assertSetEqual(t, res, []string{"CREATE", "READ"})

	o.ClearRules()

	o.LoadString("allow(_actor: User{Name: \"John\"}, _action, _resource: Widget{Id: 1});")

	actor = User{Name: "John"}
	res, err = o.AuthorizedActions(actor, resource, true)
	if err != nil {
		t.Fatalf("Failed to get allowed actions: %v", err)
	}
	if _, ok := res["*"]; !ok {
		t.Error("expected * action")
	}

	_, err = o.AuthorizedActions(actor, resource, false)
	if err == nil {
		t.Fatal("Expected an error from AuthorizedActions")
	}

	res, err = o.AuthorizedActions(actor, Widget{Id: 2}, false)
	if err != nil {
		t.Fatalf("Failed to get allowed actions: %v", err)
	}
	if len(res) != 0 {
		t.Error("expected no actions", res)
	}
}

type TestAdapter struct {
}

func (a TestAdapter) BuildQuery(filter *types.Filter) (interface{}, error) {
	return nil, nil
}

func (a TestAdapter) ExecuteQuery(query interface{}) (interface{}, error) {
	return nil, nil
}

func TestAuthorizedQuery(t *testing.T) {
	o := getOso(t)
	var err error

	o.SetDataFilteringAdapter(&TestAdapter{})

	o.LoadString("allow(_actor: User, \"get\", resource: Widget) if resource.Parent.Id = 1;")

	actor := User{Name: "Sally"}
	resource := Widget{Id: 1}
	_, _ = o.IsAllowed(actor, "get", resource)

	results, err := o.AuthorizedQuery(actor, "get", "Widget")
	fmt.Printf("%v\n", results)
	if err != nil {
		t.Fatalf("Failed to get query: %v", err)
	}
}
func TestAuthorizedFields(t *testing.T) {
	o := getOso(t)
	var res map[interface{}]struct{}

	o.LoadString( // Admins can update all fields
		"allow_field(actor: User, \"update\", _widget: Widget, field) if " +
			"  actor.Name = \"admin\" and " +
			"  field in [\"name\", \"purpose\", \"private_field\"]; " +

			// Anybody who can update a field can also read it
			"allow_field(actor, \"read\", widget: Widget, field) if " +
			"  allow_field(actor, \"update\", widget, field); " +

			// Anybody can read public fields
			"allow_field(_: User, \"read\", _: Widget, field) if " +
			"  field in [\"name\", \"purpose\"];")

	admin := User{"admin"}
	guest := User{"guest"}
	widget := Widget{0, 0}

	// Admins should be able to update all fields
	res, _ = o.AuthorizedFields(admin, "update", widget, false)
	assertSetEqual(t, res, []string{"name", "purpose", "private_field"})
	// Admins should be able to read all fields
	res, _ = o.AuthorizedFields(admin, "read", widget, false)
	assertSetEqual(t, res, []string{"name", "purpose", "private_field"})
	// Guests should not be able to update any fields
	res, _ = o.AuthorizedFields(guest, "update", widget, false)
	assertSetEqual(t, res, []string{})
	// Guests should be able to read public fields
	res, _ = o.AuthorizedFields(guest, "read", widget, false)
	assertSetEqual(t, res, []string{"name", "purpose"})
}

func TestCustomReadAction(t *testing.T) {
	var err error
	o := getOso(t)
	o.SetReadAction("fetch")

	o.LoadString("allow(\"graham\", \"fetch\", \"bar\");")

	err = o.Authorize("sam", "frob", "bar")
	// Should throw not found, sam cannot read bar
	assertAuthorizationError(t, err, true)
	err = o.Authorize("graham", "frob", "bar")
	// Should throw forbidden, graham can read bar
	assertAuthorizationError(t, err, false)
}

type CustomError struct {
	IsNotFound bool
}

func (e *CustomError) Error() string {
	return "CustomError"
}

func TestCustomErrors(t *testing.T) {
	var err error
	o := getOso(t)
	o.SetNotFoundError(func() error { return &CustomError{true} })
	o.SetForbiddenError(func() error { return &CustomError{false} })

	o.LoadString("allow(\"graham\", \"read\", \"bar\");")

	err = o.Authorize("sam", "frob", "bar")
	if custom, ok := err.(*CustomError); ok {
		if !custom.IsNotFound {
			t.Error("Expected CustomError to have IsNotFound = true")
		}
	} else {
		t.Errorf("Expected Authorize to return a CustomError, but got %v", err)
	}

	err = o.Authorize("graham", "frob", "bar")
	if custom, ok := err.(*CustomError); ok {
		if custom.IsNotFound {
			t.Error("Expected CustomError to have IsNotFound = false")
		}
	} else {
		t.Errorf("Expected Authorize to return a CustomError, but got %v", err)
	}
}

File: languages/java/oso/src/test/java/com/osohq/oso/EnforcementTest.java
package com.osohq.oso;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class EnforcementTest {
  protected Oso oso;

  public static class User {
    public String name;

    public User(String name) {
      this.name = name;
    }
  }

  public static class Widget {
    public int id;

    public Widget(int id) {
      this.id = id;
    }
  }

  public static class Request {
    public String method;
    public String path;

    public Request(String method, String path) {
      this.method = method;
      this.path = path;
    }
  }

  @BeforeEach
  public void setUp() throws Exception {
    try {
      oso = new Oso();
      oso.registerClass(User.class, "User");
      oso.registerClass(Widget.class, "Widget");
    } catch (Exception e) {
      throw new Error(e);
    }
  }

  @Test
  public void testAuthorize() throws Exception {
    User guest = new User("guest");
    User admin = new User("admin");
    Widget widget0 = new Widget(0);
    Widget widget1 = new Widget(1);

    oso.loadStr(
        "allow(_actor: User, \"read\", widget: Widget) if "
            + "widget.id = 0; "
            + "allow(actor: User, \"update\", _widget: Widget) if "
            + "actor.name = \"admin\";");

    oso.authorize(guest, "read", widget0);
    oso.authorize(admin, "update", widget1);

    // Throws a forbidden exception when user can read resource
    assertThrows(
        Exceptions.ForbiddenException.class, () -> oso.authorize(guest, "update", widget0));

    // Throws a not found exception when user cannot read resource
    assertThrows(Exceptions.NotFoundException.class, () -> oso.authorize(guest, "read", widget1));
    assertThrows(Exceptions.NotFoundException.class, () -> oso.authorize(guest, "update", widget1));

    // With checkRead = false, returns a forbidden exception
    assertThrows(
        Exceptions.ForbiddenException.class, () -> oso.authorize(guest, "read", widget1, false));
    assertThrows(
        Exceptions.ForbiddenException.class, () -> oso.authorize(guest, "update", widget1, false));
  }

  @Test
  public void testAuthorizeRequest() throws Exception {
    oso.registerClass(Request.class, "Request");
    oso.loadStr(
        "allow_request(_: User{name: \"guest\"}, request: Request) if "
            + "request.path.startsWith(\"/repos\"); "
            + "allow_request(_: User{name: \"verified\"}, request: Request) if "
            + "request.path.startsWith(\"/account\"); ");
    User guest = new User("guest");
    User verified = new User("verified");

    oso.authorizeRequest(guest, new Request("GET", "/repos/1"));
    assertThrows(
        Exceptions.ForbiddenException.class,
        () -> oso.authorizeRequest(guest, new Request("GET", "/other")));

    oso.authorizeRequest(verified, new Request("GET", "/account"));
    assertThrows(
        Exceptions.ForbiddenException.class,
        () -> oso.authorizeRequest(guest, new Request("GET", "/account")));
  }

  @Test
  public void testAuthorizedActions() throws Exception {
    oso.loadStr(
        "allow(_actor: User{name: \"sally\"}, action, _resource: Widget{id: 1})"
            + " if action in [\"CREATE\", \"READ\"];");

    User actor = new User("sally");
    Widget widget = new Widget(1);
    HashSet<Object> actions = oso.authorizedActions(actor, widget);

    assertEquals(actions.size(), 2);
    assertTrue(actions.contains("CREATE"));
    assertTrue(actions.contains("READ"));

    oso.clearRules();

    oso.loadStr(
        "allow(_actor: User{name: \"fred\"}, action, _resource: Widget{id: 2})"
            + " if action in [1, 2, 3, 4];");

    User actor2 = new User("fred");
    Widget widget2 = new Widget(2);
    HashSet<Object> actions2 = oso.authorizedActions(actor2, widget2);

    assertEquals(actions2.size(), 4);
    assertTrue(actions2.contains(1));
    assertTrue(actions2.contains(2));
    assertTrue(actions2.contains(3));
    assertTrue(actions2.contains(4));

    User actor3 = new User("doug");
    Widget widget3 = new Widget(4);
    assertTrue(oso.authorizedActions(actor3, widget3).isEmpty());
  }

  @Test
  public void testAuthorizedActionsWildcard() throws Exception {
    oso.loadStr("allow(_actor: User{name: \"John\"}, _action, _resource: Widget{id: 1});");

    User actor = new User("John");
    Widget widget = new Widget(1);

    assertEquals(Set.of("*"), oso.authorizedActions(actor, widget, true));
    assertThrows(Exceptions.OsoException.class, () -> oso.authorizedActions(actor, widget, false));
  }

  @Test
  public void testAuthorizeField() throws Exception {
    oso.loadStr(
        // Admins can update all fields
        "allow_field(actor: User, \"update\", _widget: Widget, field) if "
            + "actor.name = \"admin\" and "
            + "field in [\"name\", \"purpose\", \"private_field\"]; "
            +
            // Anybody who can update a field can also read it
            "allow_field(actor, \"read\", widget: Widget, field) if "
            + "allow_field(actor, \"update\", widget, field); "
            +
            // Anybody can read public fields
            "allow_field(_: User, \"read\", _: Widget, field) if "
            + "field in [\"name\", \"purpose\"];");
    User admin = new User("admin");
    User guest = new User("guest");
    Widget widget = new Widget(0);

    oso.authorizeField(admin, "update", widget, "purpose");
    assertThrows(
        Exceptions.ForbiddenException.class,
        () -> oso.authorizeField(admin, "update", widget, "foo"));

    oso.authorizeField(guest, "read", widget, "purpose");
    assertThrows(
        Exceptions.ForbiddenException.class,
        () -> oso.authorizeField(guest, "read", widget, "private_field"));
  }

  @Test
  public void testAuthorizedFields() throws Exception {
    oso.loadStr(
        // Admins can update all fields
        "allow_field(actor: User, \"update\", _widget: Widget, field) if "
            + "actor.name = \"admin\" and "
            + "field in [\"name\", \"purpose\", \"private_field\"]; "
            +
            // Anybody who can update a field can also read it
            "allow_field(actor, \"read\", widget: Widget, field) if "
            + "allow_field(actor, \"update\", widget, field); "
            +
            // Anybody can read public fields
            "allow_field(_: User, \"read\", _: Widget, field) if "
            + "field in [\"name\", \"purpose\"];");
    User admin = new User("admin");
    User guest = new User("guest");
    Widget widget = new Widget(0);

    // Admins should be able to update all fields
    assertEquals(
        oso.authorizedFields(admin, "update", widget),
        new HashSet(Arrays.asList("name", "purpose", "private_field")));
    // Admins should be able to read all fields
    assertEquals(
        oso.authorizedFields(admin, "read", widget),
        new HashSet(Arrays.asList("name", "purpose", "private_field")));
    // Guests should not be able to update any fields
    assertEquals(oso.authorizedFields(guest, "update", widget), new HashSet());
    // Guests should be able to read public fields
    assertEquals(
        oso.authorizedFields(guest, "read", widget), new HashSet(Arrays.asList("name", "purpose")));
  }

  @Test
  public void testCustomReadAction() throws Exception {
    oso.setReadAction("fetch");
    oso.loadStr("allow(\"graham\", \"fetch\", \"bar\");");
    assertThrows(Exceptions.NotFoundException.class, () -> oso.authorize("sam", "frob", "bar"));
    // A user who can "fetch" should get a ForbiddenException instead of a
    // NotFoundException
    assertThrows(Exceptions.ForbiddenException.class, () -> oso.authorize("graham", "frob", "bar"));
  }
}
